src/registers/model_specific.rs

```diff
@@ -172,4 +203,125 @@ mod x86_64 {
             unsafe { Self::MSR.write(address.as_u64()) };
         }
     }
+
+    impl Star {
+        /// Read the Ring 0 and Ring 3 segment bases.
+        /// The remaining fields are ignored because they are
+        /// not valid for long mode.
+        ///
+        /// # Returns
+        /// - Field 1 (SYSRET): The CS selector is set to this field + 16. SS.Sel is set to
+        /// this field + 8. Because SYSRET always returns to CPL 3, the
+        /// RPL bits 1:0 should be initialized to 11b.
+        /// - Field 2 (SYSCALL): This field is copied directly into CS.Sel. SS.Sel is set to
+        ///  this field + 8. Because SYSCALL always switches to CPL 0, the RPL bits
+        /// 33:32 should be initialized to 00b.
+        pub fn read_raw() -> (u64, u64) {
+            let msr_value = unsafe { Self::MSR.read() };
+            let sysret = msr_value.get_bits(48..63);
+            let syscall = msr_value.get_bits(32..47);
+            (sysret, syscall)
+        }
+
+        /// Read the Ring 0 and Ring 3 segment bases.
+        /// Returns
+        /// - CS Selector SYSRET
+        /// - SS Selector SYSRET
+        /// - CS Selector SYSCALL
+        /// - SS Selector SYSCALL
+        pub fn read() -> (
+            SegmentSelector,
+            SegmentSelector,
+            SegmentSelector,
+            SegmentSelector,
+        ) {
+            let raw = Self::read_raw();
+            return (
+                SegmentSelector((raw.0 + 16 - 3).try_into().unwrap()),
+                SegmentSelector((raw.0 + 8 - 3).try_into().unwrap()),
```

Comment:
Could you clarify why you're doing `- 3` here?

Reply:
The values in the registers are or'ed with 0b11 for sysret registers regardless of what the user specified. We are subtracting that back so that the user can directly match with their gdt's user mode entry rather

Reply:
The problem I see is that the user might also pass a segment selector with the requested privilege bits already set to `0b11`. In this case, it would be very strange to get a selector with requested privilege level 0 back, i.e. not the same that was written.